text/0000-const-trait-impls.md

```diff
@@ -0,0 +1,1006 @@
+This is both handled this way by the parser, and semantically what is meant here. The constness of the trait bound is affected,
+the trait bound itself exists either way.
+
+## Why do traits need to be marked as "const implementable"?
+
+### Default method bodies
+
+Adding a new method with a default body would become a breaking change unless that method/default body
+would somehow be marked as `const`, too. So by marking the trait, you're opting into the requirement that all default bodies are const checked,
+and thus neither `impl const Trait for Type` items nor `impl Trait for Type` items will be affected if you add a new method with a default body.
+This scheme avoids adding a new kind of breaking change to the Rust language,
+and instead allows everyone managing a public trait in their crate to continue relying on the
+previous rule "adding a new method is not a breaking change if it has a default body".
+
+### `~const Destruct` super trait
+
+The `Destruct` marker trait is used to name the previously unnameable drop glue that every type has.
+It has no methods, as drop glue is handled entirely by the compiler,
+but in theory drop glue could become something one can explicitly call without having to resort to extracting the drop glue function pointer from a `dyn Trait`.
+
+Traits that have `self` (by ownership) methods, will almost always drop the `self` in these methods' bodies unless they are simple wrappers that just forward to the generic parameters' bounds.
+
+The following never drops `T`, because it's the job of `<T as Add>` to handle dropping the values.
+
+```rust
+struct NewType<T>(T);
+
+impl<T: ~const Add<Output = T>> const Add for NewType<T> {
+    type Output = Self;
+    fn add(self, other: Self) -> Self::Output {
+        NewType(self.0 + other.0)
+    }
+}
+```
+
+But if any code path could drop a value...
+
+```rust
+struct NewType<T>(T, bool);
+
+struct Error;
+
+impl<T: ~const Add<Output = T>> const Add for NewType<T> {
+    type Output = Result<Self, Error>;
+    fn add(self, other: Self) -> Self::Output {
+        if self.1 {
+            Ok(NewType(self.0 + other.0, other.1))
+        } else {
+            // Drops both `self.0` and `self.1`
+            Err(Error)
+        }
+    }
+}
+```
+
+... then we need to add a `~const Destruct` bound to `T`, to ensure
+`NewType<T>` can be dropped.
+
+This bound in turn will be infectious to all generic users of `NewType` like
+
+```rust
+const fn add<T: ~const Add>(
+    a: NewType<T>,
+    b: NewType<T>,
+) -> Result<NewType<T::Output>, Error> {
+    a + b
+}
+```
+
+which now need a `T: ~const Destruct` bound, too.
+In practice we have noticed that a large portion of APIs will have a `~const Destruct` bound.
+This bound has little value as an explicit bound that appears almost everywhere.
+Especially since it is a fairly straight forward assumption that a type that has const trait impls will also have a `const Drop` impl or only contain `const Destruct` types.
```

Comment:
You do mention `const Drop` impls later inthe RFC with regards to bounds, but I'd make an additional note here or before here clarifying that `const Drop` implementations are allowed, since at first glance it's not immediately clear what their purpose would be.

A simple example might be worthwhile: even though you can't mutate global state, you could for example have a value being borrowed inside the object which gets mutated in some way in the `Drop` impl, which can be done in const context.

This also adds another interesting question: is `forget` allowed for non-`const Destruct` types? I would assume yes, but it's not clear here.

Reply:
> This also adds another interesting question: is `forget` allowed for non-`const Destruct` types? I would assume yes, but it's not clear here.

It would be a breaking change to start disallowing it, so yes, it would continue to work

Reply:
I actually had no idea what the current behaviour was for forget, so, thank you for clarifying that. Definitely worth mentioning in the RFC itself.

Reply:
I don't see a good place to put this but I also found it fairly obvious that `forget` is always fine for any input considering it has no bounds at all